nilai-api/src/nilai_api/handlers/web_search.py

```
@@ -0,0 +1,128 @@
+            system_messages_added = True
+
+    if not system_messages_added:
+        enhanced_messages.insert(0, enhanced_system_message)
+
+    return enhanced_messages, sources
+
+
+def perform_web_search_sync(query: str) -> Tuple[List[str], List[Dict[str, str]]]:
+    """
+    Synchronous version of web search for use with run_in_executor.
+
+    Args:
+        query: The search query to perform
+
+    Returns:
+        Tuple of (list of search result snippets, list of sources with metadata)
+    """
+    try:
+        with DDGS() as ddgs:
+            search_results = list(ddgs.text(query, max_results=3, region="us-en"))
+
+            results = []
+            sources = []
+
+            for result in search_results:
+                if result.get("title") and result.get("body"):
+                    title = result["title"]
+                    body = result["body"][:500]
+                    results.append(f"{title}: {body}")
+
+                    source_info = {
+                        "title": title,
+                        "url": result["href"],
+                        "snippet": body,
+                        "type": "web_search",
+                    }
+                    sources.append(source_info)
+
+            return results, sources
+
+    except Exception as e:
+        logger.error(f"Error performing web search: {e}")
+        return [], []
+
+
+async def handle_web_search(
+    req_messages: List[Message],
+) -> Tuple[List[Message], List[Dict[str, str]]]:
+    """
+    Handle web search functionality for chat requests.
+
+    Args:
+        req_messages: Original request messages
+
+    Returns:
+        Tuple of (enhanced messages with web search context if enabled, list of sources used)
+    """
+    user_query = ""
+    for message in reversed(req_messages):
+        if message.role == "user":
+            user_query = message.content
+            break
+
+    if not user_query:
+        return req_messages, []
```

This should also raise an Exception if no user query is provided and the user is notified of why the error. In this case, the status should be 4XX where the user request is malformed.